### ISSUE TITLE
Original <body> is lost when body tag is emitted due to illegal tag in <head>

### ISSUE THREAD

**Issue by apeschar** [aantron/markup.ml#54]
Consider the following document:

~~~
<!doctype html>
<html>
  <head>
    <img>
  </head>
  <body class='cls'>
  </body>
</html>
~~~

The browser parses this as containing a `body` with class `cls`. ([JSFiddle](https://jsfiddle.net/fnyqh8m6/2/))

When using Markup.ml the attributes on the `body` tag are lost because a replacement body tag seems to be generated because of the `img` in the `head`.

This is unfortunate, because in my use case the `body` element has structured data attributes that I want to extract.

I've added an incomplete start to a test case on my branch [body-attrs-after-bad-head](https://github.com/aantron/markup.ml/compare/master...apeschar:body-attrs-after-bad-head), but I don't really know how to resolve this without losing the desirable semantics of generating a body tag if it is missing altogether.

Hope you have some ideas.  Thanks for your time.

**Comment by aantron** [comment]
This is caused by a limitation of Markup.ml's design, namely that Markup.ml is a streaming parser (it deliberately does not parse the whole document before emitting signals). In contrast, HTML5 specifies a non-streaming parser that has access to the entire DOM tree throughout parsing, and can make non-local modifications to elements based on input that occurs much later than the input which created the elements. The conversion to streaming is mostly faithful, but there are a few cases (I recall three) where a correct conversion is impossible, all because they would inherently require reading the whole document in order to emit a given element correctly:

1. Nested (explicitly or due to error recovery) `<body>` elements, as in this issue. The spec [requires](https://www.w3.org/TR/html52/syntax.html#the-in-body-insertion-mode)

    > for each attribute on the token, check to see if the attribute is already present on the `body` element (the second element) on the stack of open elements, and if it is not, add the attribute and its corresponding value to that element.

2. Nested `<html>` elements, for the same kind of reason.

3. Certain bad `<table>` inputs.

This is described with much less detail in [Conformance status](https://aantron.github.io/markup.ml/#2_Conformancestatus).

My idea for addressing this was to change the parser to emit richer nesting errors, so that a user who is building a DOM can implement the non-local correction themselves. It is already possible to observe the errors emitted by the parser by using the [`?report`](https://aantron.github.io/markup.ml/#VALparse_html) argument of `parse_html`:

```ocaml
Markup.(parse_html
  ?report:(fun location e -> prerr_endline (Error.to_string ~location e))
  input
```

I think [`Error.t`](https://aantron.github.io/markup.ml/Markup.Error.html#TYPEt) can be extended to carry attribute information, which can then be included when reporting the nested `body` tag here:

https://github.com/aantron/markup.ml/blob/6d082308caa9ef01894e5d2d0485b1d16b58f2be/src/html_parser.ml#L1647-L1648

I'd probably also add a helper that can implement the spec's check

> for each attribute on the token, check to see if the attribute is already present on the `body` element (the second element) on the stack of open elements, and if it is not, add the attribute and its corresponding value to that element.

by walking along two attribute lists, but maybe this isn't so useful, as it depends on the user's DOM representation. At least, I would document what should be done in this case according to the spec (which the user can choose to ignore if they want to).

**Comment by apeschar** [comment]
Thanks for your quick and elucidating response.

I'm using Lambda Soup.  It looks like Lambda Soup is not currently using `~report`.  If I understand correctly, we could extend markup.ml to include attribute data in the error, and Lambda Soup to propagate attributes on misnested body elements to the generated body element using `~report`.

We can define `` `Misnested_tag of string * string * (string * string list) ``, or `` `Misnested_tag of Token_tag.t * string ``.  The latter would appear more sensible to me as it simplifies a lot of the code generating this error:

~~~
    | l, `Start t ->
      report l (`Misnested_tag (t, "head")) !throw mode
~~~

instead of:

~~~
    | l, `Start {name = "head"; attributes} ->
      report l (`Misnested_tag ("head", "head", attributes)) !throw mode
~~~

What do you think?

**Comment by aantron** [comment]
> we could extend markup.ml to include attribute data in the error, and Lambda Soup to propagate attributes on misnested body elements to the generated body element using `~report`.

That seems right.

> We can define `` `Misnested_tag of string * string * (string * string list) ``, or `` `Misnested_tag of Token_tag.t * string ``.

At least at first glance, I prefer the former, for two reasons:

- To avoid exposing an extra type or type abbreviation in the interface.
- `Token_tag.t`, in particular, has an additional `self_closing` field, and may have other fields that are not meant for the user to depend on. In general, it's an internal type and I'd like to retain the ability to change it at will, in principle.

As for the syntax in the source code of Markup.ml itself, it seems that we can get the shorter syntax by just defining a helper that will take the right parts out of `t` anyway, though I probably would first just use patterns like in your second example, until I became convinced that it is too annoying to do it that way :)

**Comment by aantron** [comment]
The Markup.ml part of this was addressed by @apeschar in #55. Thanks again!

**Comment by apeschar** [comment]
Closing this, problem should be fixed wherever one uses Markup.ml.  I will probably submit a PR for Lambda Soup soon.
